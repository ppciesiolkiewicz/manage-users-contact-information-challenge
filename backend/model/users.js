const faker = require("faker");
const loki = require("lokijs");
const _ = require('lodash');

const db = new loki("MSTS");
const users = db.addCollection("users", { indices: ["id"] });

const AMOUNT_OF_USERS_TO_GENERATE = 100;

const generateRandomAddress = () => ({
  country: faker.address.country(),
  streetName: faker.address.streetName(),
  streetNumber: faker.random.number().toString(),
  city: faker.address.city(),
  postalCode: faker.address.zipCode(),
  phone: faker.phone.phoneNumber(),
});

users.insert(
  _.range(AMOUNT_OF_USERS_TO_GENERATE).map(i => ({
    id: i.toString(),
    firstName: faker.name.firstName(),
    lastName: faker.name.lastName(),
    shippingAddress: generateRandomAddress(),
    billingAddress: generateRandomAddress(),
  }),
));

module.exports = {
  getById: id => users.find({ id })[0],
};
