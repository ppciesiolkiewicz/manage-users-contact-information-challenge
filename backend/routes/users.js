const express = require('express');
const router = express.Router();

const users = require('../model/users');


router.get('/:id', (req, res, next) => {
  const { params: { id } } = req;
  res.json(users.getById(id));
});

router.put('/', (req, res, next) => {
  const { body } = req;
  users.update(body);
  res.send(true);
});

module.exports = router;
