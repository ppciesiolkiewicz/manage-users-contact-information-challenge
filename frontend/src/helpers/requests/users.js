const getUserUrl = (id = '') => `http://localhost:1337/users/${id}`;

export const fetchUserData = async id => {
  return fetch(getUserUrl(id))
    .then(res => res.json());
};

export const postUserData = async user => {
  return fetch(getUserUrl(),
    {
      method: 'PUT',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(user),
    },
  )
    .then(res => res.json());
};
