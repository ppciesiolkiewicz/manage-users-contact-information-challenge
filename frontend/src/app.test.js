import React from 'react';
import renderer, { act } from 'react-test-renderer';
import wait from 'waait';
import App from './app';
import { fetchUserData as mockFetchUserData, postUserData as mockPostUserData } from './helpers/requests/users';


jest.mock('./helpers/requests/users', () => ({
  fetchUserData: jest.fn(),
  postUserData: jest.fn(),
}));

const getAddress = () => ({
  country: 'country',
  streetName: 'streetName',
  streetNumber: 'streetNumber',
  city: 'city',
  postalCode: 'postalCode',
  phone: 'phone'
});

const user = {
  id: '1',
  firstName: 'firstName',
  lastName: 'lastName',
  shippingAddress: getAddress(),
  billingAddress: getAddress(),
};

describe('App', () => {
  let rendered;

  describe('When network finishes succesfully', () => {
    beforeEach(async () => {
      mockFetchUserData.mockResolvedValue(user);
      mockPostUserData.mockResolvedValue();
      await act(async () => {
        rendered = renderer.create(<App />);
        await wait(0);
      });
    });

    it('Should render correctly', () => {
      expect(rendered.toJSON()).toMatchSnapshot();
    });
  });

  describe('When network finishes with error', () => {
    beforeEach(async () => {
      mockFetchUserData.mockRejectedValue();
      mockPostUserData.mockRejectedValue();
      await act(async () => {
        rendered = renderer.create(<App />);
        await wait(10);
      });
    });

    it('Should render error', () => {
      expect(rendered.toJSON()).toMatchSnapshot();
    });
  });

});
