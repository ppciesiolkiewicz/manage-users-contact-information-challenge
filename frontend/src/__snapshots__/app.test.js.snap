// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App When network finishes succesfully Should render correctly 1`] = `
<div
  className="app"
>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <div
      className="userDetailsContainer"
    >
      <h1>
        User details
      </h1>
      <div
        className="userDetails"
      >
        <div>
          <div
            className="input"
          >
            <label>
              First Name
            </label>
            <input
              name="firstName"
              onBlur={[Function]}
              onChange={[Function]}
              type="string"
              value="firstName"
            />
          </div>
        </div>
        <div>
          <div
            className="input"
          >
            <label>
              Last Name
            </label>
            <input
              name="lastName"
              onBlur={[Function]}
              onChange={[Function]}
              type="string"
              value="lastName"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="addressContainer"
    >
      <div
        className="address"
      >
        <h1>
          Shipping Address
        </h1>
        <div>
          <div
            className="input"
          >
            <label>
              Country
            </label>
            <input
              name="shippingAddress.country"
              onBlur={[Function]}
              onChange={[Function]}
              type="string"
              value="country"
            />
          </div>
        </div>
        <div>
          <div
            className="input"
          >
            <label>
              Street Name
            </label>
            <input
              name="shippingAddress.streetName"
              onBlur={[Function]}
              onChange={[Function]}
              type="string"
              value="streetName"
            />
          </div>
        </div>
        <div>
          <div
            className="input"
          >
            <label>
              Street number
            </label>
            <input
              name="shippingAddress.streetNumber"
              onBlur={[Function]}
              onChange={[Function]}
              type="string"
              value="streetNumber"
            />
          </div>
        </div>
        <div>
          <div
            className="input"
          >
            <label>
              City
            </label>
            <input
              name="shippingAddress.city"
              onBlur={[Function]}
              onChange={[Function]}
              type="string"
              value="city"
            />
          </div>
        </div>
        <div>
          <div
            className="input"
          >
            <label>
              Postal Code
            </label>
            <input
              name="shippingAddress.postalCode"
              onBlur={[Function]}
              onChange={[Function]}
              type="string"
              value="postalCode"
            />
          </div>
        </div>
        <div>
          <div
            className="input"
          >
            <label>
              Phone
            </label>
            <input
              name="shippingAddress.phone"
              onBlur={[Function]}
              onChange={[Function]}
              type="string"
              value="phone"
            />
          </div>
        </div>
      </div>
      <div
        className="address"
      >
        <h1>
          Billing Address
        </h1>
        <div>
          <div
            className="input"
          >
            <label>
              Country
            </label>
            <input
              name="billingAddress.country"
              onBlur={[Function]}
              onChange={[Function]}
              type="string"
              value="country"
            />
          </div>
        </div>
        <div>
          <div
            className="input"
          >
            <label>
              Street Name
            </label>
            <input
              name="billingAddress.streetName"
              onBlur={[Function]}
              onChange={[Function]}
              type="string"
              value="streetName"
            />
          </div>
        </div>
        <div>
          <div
            className="input"
          >
            <label>
              Street number
            </label>
            <input
              name="billingAddress.streetNumber"
              onBlur={[Function]}
              onChange={[Function]}
              type="string"
              value="streetNumber"
            />
          </div>
        </div>
        <div>
          <div
            className="input"
          >
            <label>
              City
            </label>
            <input
              name="billingAddress.city"
              onBlur={[Function]}
              onChange={[Function]}
              type="string"
              value="city"
            />
          </div>
        </div>
        <div>
          <div
            className="input"
          >
            <label>
              Postal Code
            </label>
            <input
              name="billingAddress.postalCode"
              onBlur={[Function]}
              onChange={[Function]}
              type="string"
              value="postalCode"
            />
          </div>
        </div>
        <div>
          <div
            className="input"
          >
            <label>
              Phone
            </label>
            <input
              name="billingAddress.phone"
              onBlur={[Function]}
              onChange={[Function]}
              type="string"
              value="phone"
            />
          </div>
        </div>
      </div>
    </div>
    <button
      type="submit"
    >
      submit
    </button>
  </form>
</div>
`;

exports[`App When network finishes with error Should render error 1`] = `
<div>
  something went wrong
</div>
`;
